### PR TITLE
NOISSUE - Add cert revocation to SDK

### LINKED ISSUE #1225: Remove Vault API client from certs service
2. Describe the feature you are requesting, as well as the possible use case(s) for it.

Vault API client brings lots of dependencies and doesnt bring much benefit in comparison to using Vault HTTP API

### LINKED ISSUE #1697: Certs PKI - vault client.RawRequest is deprecated
**FEATURE REQUEST**

1. Is there an open issue addressing this request? If it does, please add a "+1" reaction to the
   existing issue, otherwise proceed to step 2.

2. Describe the feature you are requesting, as well as the possible use case(s) for it.  

      In `certs` service, VAULT GO SDK is used for issuing and revoking certificates from VAULT. We are using function `client.RawRequest`  of SDK to sending request to VAULT. The code was present here https://github.com/mainflux/mainflux/blob/master/certs/pki/vault.go#L121   
      `RawRequest` performs the raw request given. This request may be against a Vault server not configured with this client. This is an advanced operation that generally won't need to be called externally. 
      Deprecated: This method should not be used directly. Use higher level methods instead. 
      Reference : https://pkg.go.dev/github.com/hashicorp/vault/api@v1.8.1#Client.RawRequest  
      
      As given in VAULT SDK docks , Instead of `client.RawRequest` we need to use higher level methods, 
      There is no proper documentation for how to access PKI 
      To access the VAULT API we can use the following function https://pkg.go.dev/github.com/hashicorp/vault/api#Client.Logical
      To write and read request we can we this function example : `client.Logical().Write(...)` 
      Further we also need refer example  from here : https://github.com/hashicorp/hello-vault-go
       and we need to check the VAULT CLI code working .  
       From these we can take reference and implement them in certs service
       Reference: https://github.com/hashicorp/vault/issues/15534
      
3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).  
      **must-have**

### LINKED ISSUE #1700: Revoke Certs Failing
**BUG REPORT**

1. What were you trying to achieve?
I was trying to revoke certificate the system had issued

2. What are the expected results?
To be able to get a revocation time

3. What are the received results?

From the cli tool
```bash
error: Status: : Delete "http://127.0.0.1:8204/certs/1455fb57-4e71-47fc-9201-03064cc40889": EOF
```
From the certs service
```bash
{"level":"info","message":"Method revoke_cert for token: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2NzMzNzc4MTcsImlhdCI6MTY3MzM0MTgxNywiaXNzIjoibWFpbmZsdXguYXV0aCIsInN1YiI6ImFkbWluQGV4YW1wbGUuY29tIiwiaXNzdWVyX2lkIjoiN2Y3Zjk1ZjQtNDEyYy00NTZmLTk1OGMtM2Y1M2JkYzIwM2Y0IiwidHlwZSI6MH0.oY0Qml8yuLAbpiNDzzMo0b899cIGe9M4aBx1qW_un4A and thing: 1455fb57-4e71-47fc-9201-03064cc40889 took 14.71171ms to complete without errors.","ts":"2023-01-10T16:56:09.56428127Z"}
2023/01/10 16:56:09 http: panic serving 172.18.0.1:36544: runtime error: invalid memory address or nil pointer dereference
goroutine 46 [running]:
net/http.(*conn).serve.func1()
        /usr/local/go/src/net/http/server.go:1850 +0xbf
panic({0xd78160, 0x16147a0})
        /usr/local/go/src/runtime/panic.go:890 +0x262
github.com/mainflux/mainflux/certs/pki.(*pkiAgent).Revoke(0xc0003b4b00, {0xc0000ce340, 0x3b})
        /home/r0x6f736f646f/Code/mainflux/mainflux/certs/pki/vault.go:212 +0x38e
github.com/mainflux/mainflux/certs.(*certsService).RevokeCert(0xc0002dcc40, {0x1006f48, 0xc00030ec30}, {0xc0003e8128, 0x110}, {0xc0000ce1ce, 0x24})
        /home/r0x6f736f646f/Code/mainflux/mainflux/certs/service.go:159 +0x494
github.com/mainflux/mainflux/certs/api.(*loggingMiddleware).RevokeCert(0xc000144560, {0x1006f48, 0xc00030ec30}, {0xc0003e8128, 0x110}, {0xc0000ce1ce, 0x24})
        /home/r0x6f736f646f/Code/mainflux/mainflux/certs/api/logging.go:91 +0x192
github.com/mainflux/mainflux/certs/api.(*metricsMiddleware).RevokeCert(0xc0002af890, {0x1006f48, 0xc00030ec30}, {0xc0003e8128, 0x110}, {0xc0000ce1ce, 0x24})
        /home/r0x6f736f646f/Code/mainflux/mainflux/certs/api/metrics.go:75 +0x149
github.com/mainflux/mainflux/certs/api.revokeCert.func1({0x1006f48, 0xc00030ec30}, {0xdcb1c0?, 0xc000314140?})
        /home/r0x6f736f646f/Code/mainflux/mainflux/certs/api/endpoint.go:94 +0x170
github.com/go-kit/kit/transport/http.Server.ServeHTTP({0xc00040a2a0, 0xf0c798, 0xf0c7b0, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0xc000092360, ...}, ...)
        /home/r0x6f736f646f/Code/mainflux/mainflux/vendor/github.com/go-kit/kit/transport/http/server.go:121 +0x35b
github.com/go-zoo/bone.(*Route).serveMatchedRequest(0xc000159810, {0x10063b0, 0xc0002ac1c0}, 0xc0001e0100, 0xc0003dc840?)
        /home/r0x6f736f646f/Code/mainflux/mainflux/vendor/github.com/go-zoo/bone/helper_17.go:38 +0x150
github.com/go-zoo/bone.(*Route).parse(0xc000159810, {0x10063b0, 0xc0002ac1c0}, 0xc0001e0100)
        /home/r0x6f736f646f/Code/mainflux/mainflux/vendor/github.com/go-zoo/bone/route.go:166 +0x167
github.com/go-zoo/bone.(*Mux).parse(0xc00040c140, {0x10063b0, 0xc0002ac1c0}, 0xc0001e0100)
        /home/r0x6f736f646f/Code/mainflux/mainflux/vendor/github.com/go-zoo/bone/helper.go:23 +0x99
github.com/go-zoo/bone.(*Mux).DefaultServe(0x0?, {0x10063b0, 0xc0002ac1c0}, 0x72?)
        /home/r0x6f736f646f/Code/mainflux/mainflux/vendor/github.com/go-zoo/bone/bone.go:71 +0x31
github.com/go-zoo/bone.(*Mux).ServeHTTP(0x0?, {0x10063b0?, 0xc0002ac1c0?}, 0x4675ce?)
        /home/r0x6f736f646f/Code/mainflux/mainflux/vendor/github.com/go-zoo/bone/bone.go:90 +0x96
net/http.serverHandler.ServeHTTP({0xc00030e9c0?}, {0x10063b0, 0xc0002ac1c0}, 0xc0001e0100)
        /usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.(*conn).serve(0xc000398140, {0x1006f48, 0xc000514d20})
        /usr/local/go/src/net/http/server.go:1991 +0x607
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:3102 +0x4db
```
4. What are the steps to reproduce the issue?
Using this [PR](https://github.com/mainflux/mainflux/pull/1693)
Install cli
```bash
make cli && make install
``` 
Run mainflux with certs
```bash
make run
```
```bash
cd docker
docker-compose -f addons/vault/docker-compose.yml up -d 
docker-compose -f addons/certs/docker-compose.yml up -d 
```
5. In what environment did you encounter the issue?
Using this [PR](https://github.com/mainflux/mainflux/pull/1693)

6. Additional information you deem important:
- The [response](https://github.com/mainflux/mainflux/blob/b502af65fe5d66fbc6b9c2cec5821409cb366be0/certs/pki/vault.go#L205) is empty hence s is empty

### PR DESCRIPTION
**Pull request by rodneyosodo**
### What does this do?
Fix issue certs SDK by adding BearerPrefix and changing validity to ttl

### Which issue(s) does this PR fix/relate to?
Resolves #1700.
Resolves #1697 
Closes #1225.

### List any changes that modify/break current functionality
N/A

### Have you included tests for your changes?
No

### Did you document any new/modified functionality?
No

### Notes
N/A

### CODE REVIEW

File: pkg/sdk/go/certs.go [hunk 1]
```diff
@@ -34,7 +35,7 @@ func (sdk mfSDK) IssueCert(thingID string, keyBits int, keyType, valid, token st
 	}
 
 	url := fmt.Sprintf("%s/%s", sdk.certsURL, certsEndpoint)
-	resp, err := request(http.MethodPost, token, url, d)
+	resp, err := request(http.MethodPost, apiutil.BearerPrefix+token, url, d)
```
Comment (manuio) [inline pkg/sdk/go/certs.go]: Why are you not using `sdk.processRequest` here?

Comment (rodneyosodo) [reply pkg/sdk/go/certs.go]: Updated to used `sdk.processRequest`

File: pkg/sdk/go/certs.go [hunk 2]
```diff
@@ -34,18 +37,13 @@ func (sdk mfSDK) IssueCert(thingID string, keyBits int, keyType, valid, token st
...
 	url := fmt.Sprintf("%s/%s", sdk.certsURL, certsEndpoint)
-	resp, err := request(http.MethodPost, token, url, d)
-	if err != nil {
-		return Cert{}, errors.NewSDKError(err)
-	}
-	defer resp.Body.Close()
-
-	if err := errors.CheckError(resp, http.StatusOK); err != nil {
-		return Cert{}, err
+	_, body, sdkerr := sdk.processRequest(http.MethodPost, url, token, string(CTJSON), d, http.StatusCreated)
```
Comment (manuio) [inline pkg/sdk/go/certs.go:37]: You should remove `request` method and replace it everywhere by `sdk.processRequest`

**Review by manuio — CHANGES_REQUESTED** [review]

File: pkg/sdk/go/certs.go (hunk 2 above)
Comment (rodneyosodo) [reply pkg/sdk/go/certs.go:37]: On it

**Review by drasko — DISMISSED** [review]
LGTM

File: certs/api/responses.go [hunk 3]
```diff
@@ -28,6 +28,10 @@ type certsRes struct {
 	created    bool
 }
 
+type revokeCertsRes struct {
+	RevocationTime time.Time `mapstructure:"revocation_time"`
```
Comment (arvindh123) [inline certs/api/responses.go]: Do we need the `mapstructure` tag ?

File: certs/pki/vault.go [hunk 4]
```diff
@@ -179,15 +148,18 @@ func (p *pkiAgent) Read(serial string) (Cert, error) {
 }
 
 func (p *pkiAgent) Revoke(serial string) (time.Time, error) {
+	// We aren't using p.client.Logical().Delete() since it is making a DELETE request to vault
+	// yet for certificate revoking vault needs a POST request to the same URL
+	// This is a library issue and will be solved once a POST request is added to p.client.Logical()
+	// as the current alternatives either use a PUT or DELETE request
 	cReq := certRevokeReq{
-		SerialNumber: serial,
+		SerialNumber: strings.Replace(serial, ":", "-", -1),
```
Comment (dborovcanin) [inline certs/pki/vault.go]: Can you add a quick comment explaining this line?

File: certs/api/responses.go (hunk 3 above)
Comment (rodneyosodo) [reply certs/api/responses.go]: Fixed

File: certs/pki/vault.go (hunk 4 above)
Comment (dborovcanin) [reply certs/pki/vault.go]: Use `http.MethodPost` instead of the string in:
```go
r := p.client.NewRequest("POST", p.revokeURL)
```

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: certs/pki/vault.go (hunk 4 above)
Comment (rodneyosodo) [reply certs/pki/vault.go]: updated

**Review by drasko — DISMISSED** [review]
LGTM

**Comment by rodneyosodo** [comment]
Removing `key_bits` and `key_type` during issue cert as it raises this error `"Endpoint ignored these unrecognized parameters: [key_bits key_type]"`

[Ref](https://developer.hashicorp.com/vault/tutorials/secrets-management/pki-engine#step-4-request-certificates)

File: certs/mocks/pki.go [hunk 5]
```diff
@@ -21,6 +21,10 @@ import (
 	"github.com/mainflux/mainflux/pkg/errors"
 )
 
+const (
+	keyBits = 2048
+)
```
Comment (dborovcanin) [inline certs/mocks/pki.go]: Inline `const keyBits = 2048`.

File: docker/addons/vault/vault-init.sh [hunk 6]
```diff
@@ -1,10 +1,21 @@
...
+}
 vault() {
     docker exec -it mainflux-vault vault "$@"
 }
 
 mkdir -p data
 
 vault operator init 2>&1 | tee >(sed -r 's/\x1b\[[0-9;]*m//g' > data/secrets)
+
+write_env
```
Comment (dborovcanin) [inline docker/addons/vault/vault-init.sh]: Add an empty line.

File: provision/config.go [hunk 7]
```diff
@@ -51,9 +51,9 @@ type Gateway struct {
 
 // Cert represetns the certificate config
 type Cert struct {
-	TTL     string `json:"ttl" toml:"ttl"`
-	KeyBits int    `json:"key_bits" toml:"key_bits"`
-	KeyType string `json:"key_type"`
+	TTL string `json:"ttl" toml:"ttl"`
+	// KeyBits int    `json:"key_bits" toml:"key_bits"`
+	// KeyType string `json:"key_type"`
```
Comment (dborovcanin) [inline provision/config.go]: Remove commented code.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: provision/service.go [hunk 8]
```diff
@@ -263,7 +263,7 @@ func (ps *provisionService) Cert(token, thingID, ttl string, keyBits int) (strin
 	if err != nil {
```
Comment (arvindh123) [inline provision/service.go:262]: Remove the unused parameter  `keyBits` from the function

File: docker/addons/vault/vault-init.sh [hunk 9]
```diff
@@ -1,10 +1,21 @@
...
 
+scriptdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+export MAINFLUX_DIR=$scriptdir/../../../
+
+write_env() {
+    sed -i "s,MF_VAULT_UNSEAL_KEY_1=.*,MF_VAULT_UNSEAL_KEY_1=$(awk -F ": " '$1 == "Unseal Key 1" {print $2}' data/secrets)," $MAINFLUX_DIR/docker/.env
+    sed -i "s,MF_VAULT_UNSEAL_KEY_2=.*,MF_VAULT_UNSEAL_KEY_2=$(awk -F ": " '$1 == "Unseal Key 2" {print $2}' data/secrets)," $MAINFLUX_DIR/docker/.env
+    sed -i "s,MF_VAULT_UNSEAL_KEY_3=.*,MF_VAULT_UNSEAL_KEY_3=$(awk -F ": " '$1 == "Unseal Key 3" {print $2}' data/secrets)," $MAINFLUX_DIR/docker/.env
+    sed -i "s,MF_VAULT_TOKEN=.*,MF_VAULT_TOKEN=$(awk -F ": " '$1 == "Initial Root Token" {print $2}' data/secrets)," $MAINFLUX_DIR/docker/.env
+}
```
Comment (dborovcanin) [inline docker/addons/vault/vault-init.sh:12]: I like this one. :)

**Review by dborovcanin — APPROVED** [review]

File: docker/addons/vault/vault-init.sh (hunk 9 above)
Comment (arvindh123) [reply docker/addons/vault/vault-init.sh:12]: +1

**Review by drasko — APPROVED** [review]
LGTM

File: certs/pki/vault.go [hunk 10]
```diff
@@ -98,44 +91,29 @@ func NewVaultClient(token, host, path, role string) (Agent, error) {
...
-
-	r := p.client.NewRequest("POST", p.issueURL)
-	if err := r.SetJSONBody(cReq); err != nil {
-		return Cert{}, err
-	}
-
-	resp, err := p.client.RawRequest(r)
-	if resp != nil {
-		defer resp.Body.Close()
+		TTL:        fmt.Sprintf("%sh", ttl),
```
Comment (arvindh123) [inline certs/pki/vault.go:104]: Do we need to add `h` prefix to ttl ?
Why the request can we with any prefix like `h`,`m`, `s`

**Pull request merged**